api/routes/apds/activities/goals/put.js

```diff
@@ -0,0 +1,98 @@
+const logger = require('../../../../logger')('apd activites route put');
+const {
+  apdActivity: defaultActivityModel,
+  apdActivityGoal: defaultGoalModel,
+  apdActivityGoalObjective: defaultObjectiveModel
+} = require('../../../../db').models;
+const { userCanEditAPD: defaultUserCanEditAPD } = require('../../utils');
+const loggedIn = require('../../../../auth/middleware').loggedIn;
+
+module.exports = (
+  app,
+  ActivityModel = defaultActivityModel,
+  GoalModel = defaultGoalModel,
+  ObjectiveModel = defaultObjectiveModel,
+  userCanEditAPD = defaultUserCanEditAPD
+) => {
+  logger.silly('setting up PUT /activities/:id/goals route');
+  app.put('/activities/:id/goals', loggedIn, async (req, res) => {
+    logger.silly(req, 'handling PUT /activities/:id/goals route');
+    logger.silly(
+      req,
+      `attempting to update goals on activity [${req.params.id}]`
+    );
+
+    try {
+      const activityID = +req.params.id;
+      const activity = await ActivityModel.where({ id: activityID }).fetch({
+        withRelated: ['apd', 'goals.objectives']
+      });
+      if (!activity) {
+        logger.verbose(req, 'activity not found');
+        return res.status(404).end();
+      }
+      const apdID = activity.related('apd').get('id');
+
+      if (!await userCanEditAPD(req.user.id, apdID)) {
+        logger.verbose(
+          req,
+          'user (state) not associated with the apd this activity belongs to'
+        );
+        return res.status(404).end();
+      }
+
+      if (!Array.isArray(req.body)) {
+        logger.verbose('request is not an array');
+        return res
+          .status(400)
+          .send({ error: 'edit-activity-invalid-goals' })
+          .end();
+      }
+
+      // Delete the previous goals and objectives for this activity
+      logger.silly('deleting previous goals and objectives for this activity');
+      activity.related('goals').forEach(async goal => {
+        goal.related('objectives').forEach(async objective => {
+          await objective.destroy();
+        });
+        await goal.destroy();
+      });
+
+      const awaiting = req.body.map(async goal => {
+        if (goal.description) {
+          const goalModel = GoalModel.forge({
+            description: goal.description,
+            activity_id: activityID
+          });
+          await goalModel.save();
+          const goalID = goalModel.get('id');
+
+          if (Array.isArray(goal.objectives)) {
+            goal.objectives.forEach(async objective => {
+              if (typeof objective === 'string' && objective) {
+                const objectiveModel = ObjectiveModel.forge({
```

Comment:
future optimization maybe: adding a helper method on `ObjectiveModel` to add objects; something like:

`addObjectives(objs, goal) { `await this.insert(objs.map(o => ({ description: o, activity_goal_id: goal }))`